File: 70.爬楼梯.py
#
# @lc app=leetcode.cn id=70 lang=python3
#
# [70] 爬楼梯
#

# @lc code=start
class Solution:
    def climbStairs(self, n: int) -> int:
        # 每次其实都只要考虑最后一步是跨一个还是两个，那么就是f(x-1)+f(x-2)
        a = 0
        b = 1
        c = 0
        for i in range(n):
            c = a + b
            a = b
            b = c
        return c
# @lc code=end


if __name__ == "__main__":
    s = Solution()
    s.climbStairs(1)
    s.climbStairs(2)
    s.climbStairs(3)
    s.climbStairs(4)
    s.climbStairs(5)

File: 70.爬楼梯.cpp
/*
 * @lc app=leetcode.cn id=70 lang=cpp
 *
 * [70] 爬楼梯
 */
#include "commom.hpp"
// @lc code=start
class Solution {
  //  private:
  //   int cnt = 0;

 public:
  int climbStairs(int n) {
    // dfs(n);
    vector<int> dp(3);
    dp[0] = 1;
    dp[1] = 2;
    if (n <= 2) {
      return dp[n - 1];
    }

    for (size_t i = 2; i < n; i++) {
      dp[2] = dp[0] + dp[1];
      dp[0] = dp[1];
      dp[1] = dp[2];
    }
    return dp[2];
  }

  // void dfs(int n) {
  //   if (n == 0) {
  //     cnt++;
  //     return;
  //   }
  //   /* do choice */

  //   if (n == 1) {
  //     // if only 1 step, choice step 1
  //     dfs(n - 1);
  //   } else {
  //     // else choice 1 or 2
  //     dfs(n - 1);
  //     dfs(n - 2);
  //   }
  // }
};
// @lc code=end

int main(int argc, char const* argv[]) {
  Solution s;
  ic(s.climbStairs(1));
  return 0;
}
